pkg/kubectl/cmd/config/navigation_step_parser.go

```diff
@@ -0,0 +1,226 @@
+/*
+Copyright 2014 Google Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package config
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+
+	clientcmdapi "github.com/GoogleCloudPlatform/kubernetes/pkg/client/clientcmd/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+)
+
+type navigationSteps struct {
+	steps            []navigationStep
+	currentStepIndex int
+}
+
+type navigationStep struct {
+	stepValue string
+	stepType  stepLocation
+}
+
+type stepLocation string
+
+const (
+	done            = stepLocation("done")
+	root            = stepLocation("root")
+	rootValue       = stepLocation("root-value")
+	preference      = stepLocation("preference")
+	preferenceValue = stepLocation("preference-value")
+	cluster         = stepLocation("cluster")
+	clusterName     = stepLocation("cluster-name")
+	clusterValue    = stepLocation("cluster-value")
+	context         = stepLocation("context")
+	contextName     = stepLocation("context-name")
+	contextValue    = stepLocation("context-value")
+	authInfo        = stepLocation("authInfo")
+	authInfoName    = stepLocation("authInfo-name")
+	authInfoValue   = stepLocation("authInfo-value")
+)
+
+var (
+	configValueOptions     = getPotentialTypeValues(reflect.ValueOf(clientcmdapi.NewConfig()))
+	preferenceValueOptions = getPotentialTypeValues(reflect.ValueOf(clientcmdapi.NewPreferences()))
+	clusterValueOptions    = getPotentialTypeValues(reflect.ValueOf(clientcmdapi.NewCluster()))
+	contextValueOptions    = getPotentialTypeValues(reflect.ValueOf(clientcmdapi.NewContext()))
+	authInfoValueOptions   = getPotentialTypeValues(reflect.ValueOf(clientcmdapi.NewAuthInfo()))
+)
+
+func newNavigationSteps(path string) (*navigationSteps, error) {
+	steps := []navigationStep{}
+	individualParts := strings.Split(path, ".")
+
+	currType := root
+	currPartIndex := 0
+	for currPartIndex < len(individualParts) {
+		switch currType {
+		case root:
+			nextPart := individualParts[currPartIndex]
+			nextType := rootValue
+			switch nextPart {
+			case "clusters":
+				nextType = cluster
+			case "users":
+				nextType = authInfo
+			case "contexts":
+				nextType = context
+			case "preferences":
+				nextType = preference
+			}
+
+			if !configValueOptions.Has(nextPart) {
+				return nil, fmt.Errorf("unable to parse %v after %v at %v", path, steps, currType)
+			}
+
+			steps = append(steps, navigationStep{nextPart, nextType})
+			currPartIndex += 1
+			currType = nextType
+
+		case cluster:
```

Comment:
> yuck - anyway to dry this up?

Metadata map or reflection (blech).  Since the metadata can be built using reflection and that addresses your other concern, I'll do it like that.